### PR TITLE
Fix: RMSNormalization crash on torch backend

### PR DESCRIPTION
**Pull request by maitry63**
This PR resolves an issue where RMSNormalization would trigger a RuntimeError on the Torch backend.

Fixes : [22201](https://github.com/keras-team/keras/issues/22201)

### CODE REVIEW

File: keras/src/layers/normalization/rms_normalization.py [hunk 1]
```diff
@@ -48,31 +48,41 @@ def __init__(self, axis=-1, epsilon=1e-6, **kwargs):
...
-        if isinstance(self.axis, list):
-            shape = tuple([input_shape[dim] for dim in self.axis])
-        else:
-            shape = (input_shape[self.axis],)
-            self.axis = [self.axis]
+        ndim = len(input_shape)
+        axes = (
+            self.axis if isinstance(self.axis, (list, tuple)) else [self.axis]
+        )
+        axes = [axis + ndim if axis < 0 else axis for axis in axes]
```
Comment (hertschuh) [inline keras/src/layers/normalization/rms_normalization.py]: Please import and use `canonicalize_axis`:

https://github.com/keras-team/keras/blob/master/keras/src/backend/common/backend_utils.py#L280

It adds some validation too.

File: keras/src/layers/normalization/rms_normalization.py [hunk 2]
```diff
@@ -48,31 +48,41 @@ def __init__(self, axis=-1, epsilon=1e-6, **kwargs):
...
+        self.axis = sorted(list(set(axes)))
+        shape = tuple([input_shape[dim] for dim in self.axis])
```
Comment (hertschuh) [inline keras/src/layers/normalization/rms_normalization.py]: nitpick: no need for the square brackets:

```python
shape = tuple(input_shape[dim] for dim in self.axis)
```

File: keras/src/layers/normalization/rms_normalization.py [hunk 3]
```diff
@@ -48,31 +48,41 @@ def __init__(self, axis=-1, epsilon=1e-6, **kwargs):
...
         """
-        return ops.rms_normalization(
-            x, scale=self.scale, axis=self.axis, epsilon=self.epsilon
-        )
+        x = ops.cast(x, self.compute_dtype)
+        pow_2 = ops.square(x)
+        ms = ops.mean(pow_2, axis=self.axis, keepdims=True)
+        rms = ops.sqrt(ms + self.epsilon)
+        normalized = x / rms
+        return normalized * self.scale
```
Comment (hertschuh) [inline keras/src/layers/normalization/rms_normalization.py]: Why was this changed?

File: keras/src/layers/normalization/rms_normalization.py [hunk 4]
```diff
@@ -48,31 +48,41 @@ def __init__(self, axis=-1, epsilon=1e-6, **kwargs):
...
+        )
+        axes = [axis + ndim if axis < 0 else axis for axis in axes]
+        self.axis = sorted(list(set(axes)))
+        shape = tuple([input_shape[dim] for dim in self.axis])
 
         self.scale = self.add_weight(
-            name="scale", shape=shape, initializer="ones"
+            name="scale", shape=shape, initializer="ones", trainable=True
         )
-
```
Comment (hertschuh) [inline keras/src/layers/normalization/rms_normalization.py:60]: `trainable=True` is implied BTW, it's the default.

File: keras/src/layers/normalization/rms_normalization.py (hunk 1 above)
Comment (maitry63) [reply keras/src/layers/normalization/rms_normalization.py]: Thank you  so much for review ,  done. I have replaced the manual axis handling with `canonicalize_axis`. Implemented this via a list comprehension to correctly handle `multi-axis inputs` as well.

File: keras/src/layers/normalization/rms_normalization.py (hunk 2 above)
Comment (maitry63) [reply keras/src/layers/normalization/rms_normalization.py]: Corrected.

File: keras/src/layers/normalization/rms_normalization.py (hunk 3 above)
Comment (maitry63) [reply keras/src/layers/normalization/rms_normalization.py]: Here,  actually the decomposition into `square,` `mean`, and `sqrt` is added for PyTorch backend stability. 
Previous, torch implementation of `RMSnormalization` often crashes when encountering specific reduction axes. 
So, used explicit ops to ensure the layer is robust across all backends  without functional regressions.

File: keras/src/layers/normalization/rms_normalization.py (hunk 4 above)
Comment (maitry63) [reply keras/src/layers/normalization/rms_normalization.py:60]: Yes, thank you, updated. Removed the `trainable=True` argument from `add_weight` as it is implied by default & make the code clean.

**Comment by maitry63** [comment]
Addressed the above comments.

File: keras/src/layers/normalization/rms_normalization.py (hunk 3 above)
Comment (hertschuh) [reply keras/src/layers/normalization/rms_normalization.py]: So, if that's the case, the op implementation itself should be fixed instead:

https://github.com/keras-team/keras/blob/21f435bcebe9d8f85b14bb6e687ce11c80602f19/keras/src/ops/nn.py#L3001-L3007

And the layer code should not be modified.

So what you're saying is that the optimization that uses `torch.nn.functional.rms_norm` doesn't work?

Comment (maitry63) [reply keras/src/layers/normalization/rms_normalization.py]: Yes, thank you for pointing out , you're correct, the underlying build and functional logic of the `rms_normalization `shouldn't be changed to ensure consistency. I've only updated the implementation to use PyTorch's native multi-dimensional mean instead of a manual loop.

File: keras/src/backend/torch/nn.py [hunk 5]
```diff
@@ -966,6 +949,26 @@ def batch_normalization(
     )
 
 
+def rms_normalization(x, scale, axis, epsilon):
```
Comment (hertschuh) [inline keras/src/backend/torch/nn.py:952]: This will never be called. There is nothing connecting to this code.

This is the current code path: https://github.com/keras-team/keras/blob/0d5c2f778d130ac1df79bee25ddd941b551519e9/keras/src/ops/nn.py#L2988

**Review by hertschuh — COMMENTED** [review]
Thanks for addressing this.

The rest looks good, but you don't need the changes to `keras/src/backend/torch/nn.py`.

**Comment by hertschuh** [comment]
This was addressed in https://github.com/keras-team/keras/pull/22296
